### PR TITLE
[KIWI-1470]BE | Handle new IPV Core flag for the No-Photo-ID Route

### PR DESCRIPTION
**Pull request by hemavana**
### What changed

Storing the journey type for a session based on the context claim set by IPV core.
Also created a new /session-config api endpoint for FE to fetch the journey_type for a session
Evidences attached in the ticket for reference.
### Why did it change

To handle the BAV (No-Photo-ID route) from FE

### Issue tracking

- [KIWI-1470](https://govukverify.atlassian.net/browse/KIWI-1470)

### PII logging

- [x] Verified that no PII data is being logged

### Environment variables or secrets

[KIWI-1470]: https://govukverify.atlassian.net/browse/KIWI-1470?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: deploy/cic-spec.yaml [hunk 1]
```diff
@@ -580,6 +580,93 @@ paths:
...
+    get:
+      operationId: getSessionConfig
+      summary: Return journey type for a session
+      tags:
+        - Backend - Common CRI Lambda
+      parameters:
+        - $ref: "#/components/parameters/SessionHeader"
+      responses:
+        "200":
+          description: "Return journy type"
```
Comment (Caitlin-cooling) [inline deploy/cic-spec.yaml:594]: Typo in `journy`

File: deploy/cic-spec.yaml [hunk 2]
```diff
@@ -987,6 +1074,14 @@ components:
...
             - CIC_AUTH_CODE_ISSUED
             - CIC_ACCESS_TOKEN_ISSUED
+    SessionConfigResponse:
+      required:
+        - "journey_type"
+      type: "object"
+      properties:
+        journey_type:
+          type: "string"
+          example: "face-to-face"
```
Comment (Caitlin-cooling) [inline deploy/cic-spec.yaml:1084]: Do you think we should have an enum for this? If it's going to control frontend routing it's pretty important that it's correct I think

File: src/services/SessionConfigRequestProcessor.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+			SessionConfigRequestProcessor.instance = new SessionConfigRequestProcessor(logger, metrics);
+		}
+		return SessionConfigRequestProcessor.instance;
+	}
+
+	async processRequest(sessionId: string): Promise<Response> {
+
+		const session = await this.cicService.getSessionById(sessionId);
+
+		if (session != null) {
```
Comment (Caitlin-cooling) [inline src/services/SessionConfigRequestProcessor.ts:45]: We have a couple of specific null checks in this file, is there a reason that we've gone for this over implicitly checking for truthy values that would catch `undefined` as well?

File: src/utils/Constants.ts [hunk 4]
```diff
@@ -53,4 +55,10 @@ export class Constants {
     static readonly GIVEN_NAME_REGEX = /^[a-zA-Z.'-]+( [a-zA-Z.'-]+)*$/;
 
 	static readonly TXMA_FIELDS_TO_SHOW = ["event_name", "session_id", "govuk_signin_journey_id"];
+
+    static readonly FACE_TO_FACE_JOURNEY = "Face-to-Face";
+
+    static readonly NO_PHOTO_ID_JOURNEY = "no-photo-id";
```
Comment (Caitlin-cooling) [inline src/utils/Constants.ts:61]: [nit] do you think we can align the casing between the two journeys?

File: src/utils/ValidationHelper.ts [hunk 5]
```diff
@@ -94,6 +94,8 @@ export class ValidationHelper {
 			return `JWT validation/verification failed: Unable to retrieve redirect URI for client_id: ${requestBodyClientId}`;
 		} else if (expectedRedirectUri !== jwtPayload.redirect_uri) {
 			return `JWT validation/verification failed: Redirect uri ${jwtPayload.redirect_uri} does not match configuration uri ${expectedRedirectUri}`;
+		} else if (jwtPayload.context != null && expectedContext !== jwtPayload.context) {
```
Comment (Caitlin-cooling) [inline src/utils/ValidationHelper.ts:97]: Sorry I'm a little confused about this, would you be able to elaborate? Are we saying where that we are rejecting any JWT where the context provided is not Constants.EXPECTED_CONTEXT (bank_account) ?

File: src/services/SessionConfigRequestProcessor.ts (hunk 3 above)
Comment (hemavana) [reply src/services/SessionConfigRequestProcessor.ts:45]: Which other checks are you referring to? in here we are checking if we get a session object for the given session-id else throw an error.

File: deploy/cic-spec.yaml [hunk 6]
```diff
@@ -580,6 +580,93 @@ paths:
...
         type: "aws"
+  
+  /session-config:
+    get:
+      operationId: getSessionConfig
+      summary: Return journey type for a session
+      tags:
+        - Backend - Common CRI Lambda
+      parameters:
+        - $ref: "#/components/parameters/SessionHeader"
```
Comment (Caitlin-cooling) [inline deploy/cic-spec.yaml:591]: I see you've changed everything to `x-govuk-signin-session-id`, I think this also needs to change to `SessionId`

File: src/services/SessionConfigRequestProcessor.ts (hunk 3 above)
Comment (Caitlin-cooling) [reply src/services/SessionConfigRequestProcessor.ts:45]: Referring to this `session != null`, and on line 51 `session?.journey != null`. 

I see why are you checking for `null` but I'm wondering why we aren't checking for all falsy values by using `if (session)` and `session?.journey ?? Constants.FACE_TO_FACE_JOURNEY`

File: src/utils/ValidationHelper.ts (hunk 5 above)
Comment (Caitlin-cooling) [reply src/utils/ValidationHelper.ts:97]: Why is it considered an invalid request? I see in the mock we can attach anything we like

Comment (hemavana) [reply src/utils/ValidationHelper.ts:97]: I think IPV core may add anything in the context and hence the mock is doing that as per 1472 ticket. CIC needs to reject any such requests from IPV where context is not bank_account as we are supporting only this type for now.

File: src/services/SessionConfigRequestProcessor.ts (hunk 3 above)
Comment (hemavana) [reply src/services/SessionConfigRequestProcessor.ts]: Done, Thanks.

File: deploy/cic-spec.yaml (hunk 6 above)
Comment (hemavana) [reply deploy/cic-spec.yaml]: Good catch. Thanks

File: src/utils/ValidationHelper.ts (hunk 5 above)
Comment (Caitlin-cooling) [reply src/utils/ValidationHelper.ts:97]: Ah I see thanks for explaining!

**Review by Caitlin-cooling — DISMISSED** [review]
Nice stuff!

**Review by Caitlin-cooling — DISMISSED** [review]

**Review by Caitlin-cooling — APPROVED** [review]

**Pull request merged**